File: problem-935-stamping-the-sequence.cpp
// trying to check if the performance stats are per language or not
// the exact port of the python solution
#include <string>
#include <iostream>

using namespace std;

class Solution {
public:
    vector<int> movesToStamp(string stamp, string target) {
        int n = target.length();
        int k = stamp.length();

        vector<bool> matched(n, false);

        vector<int> result;
        int matchedCount = 0;

        while (matchedCount != n) {
            bool progressed = false;

            for (int offset = 0; offset <= n - k; offset++) {
                // std::cout << offset << std::endl;
                bool isMatch = true;
                int newMatches = 0;

                for (int idx = 0; idx < k; idx++) {
                    if (stamp[idx] != target[offset + idx] && !matched[offset + idx]) {
                        isMatch = false;
                        break;
                    }

                    if (matched[offset + idx] != true) {
                        newMatches += 1;
                    }
                }

                if (newMatches != 0 && isMatch)
                {
                    for (int idx = 0; idx < k; idx++) {
                        matched[idx + offset] = true;
                    }
                    matchedCount += newMatches;
                    result.push_back(offset);
                    progressed = true;
                    break;
                }
            }

            if (!progressed && matchedCount != n) {
                result = vector<int>();
                // std::cout << "boom" << " " << matchedCount;
                break;
            }
        }

        reverse(result.begin(), result.end());

        return result;
    }
};

template <typename T, typename A>
void printVector(vector<T, A> const &vector)
{
    std::cout << "[ ";
    for (auto x: vector) {
        std::cout << x << " ";
    }
    std::cout << "]" << std::endl;
}

int main() {
    Solution x;

    printVector(x.movesToStamp("abc", "ababc"));

    return 0;
}

File: problem-935-stamping-the-sequence.py
from typing import List


class Solution:
    def movesToStamp(self, stamp: str, target: str) -> List[int]:

        n, k = len(target), len(stamp)
        matched = [False] * n

        def search(matched_count: int) -> List[int]:

            # try to match stamp completely
            for offset in range(0, n - k + 1):
                is_match = True
                new_matches_count = 0

                for idx in range(k):

                    if stamp[idx] != target[offset + idx] and not matched[offset + idx]:
                        is_match = False
                        break

                    if not matched[offset + idx]:
                        new_matches_count += 1

                if new_matches_count and is_match:
                    for idx in range(k):
                        matched[idx + offset] = True

                    inner_result = search(matched_count + new_matches_count)

                    if not inner_result and (matched_count + new_matches_count) != n:
                        return []

                    return [offset] + inner_result

            return []

        return list(reversed(search(0)))


if __name__ == '__main__':
    x = Solution()
    print(x.movesToStamp('abc', 'ababc'))
    print(x.movesToStamp("abca", "aabcaca"))
    print(x.movesToStamp('oz', 'ooozz'))
    print(x.movesToStamp('cab', 'cabbb'))
    print(x.movesToStamp('oz', 'ooz'))
    print(x.movesToStamp('mda', 'mdadddaaaa'))
